Extract maven dependencies out of app-engine into optional maven module

etanshaul opened the pull request:
fixes #2183 and #1939 

Creates a new, optionally included, app-engine-maven module that contains all of the maven dependencies related to app-engine.

The app-engine module has no direct maven dependencies. All of the dependencies are inverted via the extension mechanism done in previous PRs.

With this change, all App Engine functionality works as expected _even if the maven plugin is uninstalled by the user_. In this case, only the maven specific app-engine features will be unavailable (for example, the maven-based deployment sources will be unavailable). This also fixes the Maven related `NoClassDefFound` errors we were seeing.

The sets us up closer to using app-engine in other IDEs as well.

etanshaul commented:
maven plugin installed:
![image](https://user-images.githubusercontent.com/1735744/41796931-36d5c1be-7636-11e8-9125-f52bfc7658cf.png)

both sources including maven now available (maven source comes from new maven module):
![image](https://user-images.githubusercontent.com/1735744/41796963-4b226da2-7636-11e8-9305-1a350575ddec.png)

----

maven plugin not installed:
![image](https://user-images.githubusercontent.com/1735744/41796858-f2cb9fa2-7635-11e8-9cb7-1a2f97fa0d1f.png)

no maven deployment sources available:
![image](https://user-images.githubusercontent.com/1735744/41796871-02946ac2-7636-11e8-8206-fd05960af868.png)

previously, we'd just get a bunch of `NoClassDefFound` errors when trying to use the App Engine functionality when the Maven plugin was not installed.

etanshaul commented on app-engine/java/gradle/build.gradle:
```diff
@@ -22,7 +22,7 @@ sourceSets.test.java.srcDirs = ['testSrc']
 sourceSets.test.resources.srcDirs = ['testResources']
 
 intellij {
-    plugins 'Groovy', 'gradle', 'git4idea', 'properties', 'junit', 'yaml', 'maven'
```
this was just some cleanup I noticed along the way

etanshaul commented:
I need to do a bit of test refactoring here too to get the build to pass. I had a test dependency in the app engine module on maven which I need to cut.

patflynn commented on google-cloud-tools-plugin/resources/META-INF/plugin.xml:
```diff
@@ -50,6 +50,7 @@
   <xi:include href="/META-INF/google-cloud-core.xml" xpointer="xpointer(/idea-plugin/*)"/>
   <xi:include href="/META-INF/google-cloud-storage.xml" xpointer="xpointer(/idea-plugin/*)"/>
   <xi:include href="/META-INF/google-cloud-repos.xml" xpointer="xpointer(/idea-plugin/*)"/>
+  <xi:include href="/META-INF/google-cloud-maven.xml" xpointer="xpointer(/idea-plugin/*)"/>
```
I think you want to invert this so that this import is optional and the google-cloud-maven.xml uses xi:include.  Once you decouple libraries support from maven you'll probably want to also include a libraries maven plugin fragment into google-cloud-maven.xml but you can only have one optional inclusion per module type.

etanshaul replied:
ah yeah. thanks. I initially tried having an optional depends here, and then a regular depends in "google-cloud-maven". but that clearly wont work.

etanshaul commented:
@ivanporty @patflynn should be good to go

changes:
- per @patflynn point, I inverted the maven dependency inclusion so that `google-cloud-maven.xml`  is imported via an optional depends. `google-cloud-maven.xml` then xi:include's all of the maven dependent modules (which now is app-engine-java-gradle, and google-cloud-apis). see https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/2192/commits/0bd0935d2d155ac99af32bd65c4d072a32688410
- fixed tests
- refactored MavenTestUtils so that `MavenTestUtils#runWithMavenModule` handles all of the setup / teardown of creating a maven module. This required some changes to tests unrelated to this pr...sorry about that..

ivanporty commented on app-engine/java/maven/resources/META-INF/app-engine-java-maven.xml:
```diff
@@ -0,0 +1,33 @@
+        <appEngineStandardLibraryManager
+                implementation="com.google.cloud.tools.intellij.appengine.java.maven.AppEngineStandardMavenLibraryManager"/>
+    </extensions>
+</idea-plugin>
```
Should we end this with an empty line?

etanshaul replied:
done

ivanporty commented on google-cloud-tools-plugin/resources/META-INF/plugin.xml:
```diff
@@ -46,7 +46,7 @@
   <depends optional="true" config-file="javaee-integration.xml">com.intellij.javaee</depends>
   <depends optional="true" config-file="gwt-integration.xml">com.intellij.gwt</depends>
-  <depends optional="true" config-file="google-cloud-apis.xml">org.jetbrains.idea.maven</depends>
+  <depends optional="true" config-file="google-cloud-maven.xml">org.jetbrains.idea.maven</depends>
```
Was this intentional or we accidentally lost cloud API module dependency?

ivanporty replied:
ah I got it so this module intention is to be an umbrella module for all maven related sub-modules, correct? and this (for now) includes the whole cloud apis module?

etanshaul replied:
yes exactly (to your 2nd comment)

ivanporty commented on google-cloud-tools-plugin/resources/META-INF/google-cloud-maven.xml:
```diff
@@ -0,0 +1,20 @@
+  ~ limitations under the License.
+  -->
+
+<idea-plugin xmlns:xi="http://www.w3.org/2001/XInclude">
```
probably makes sense to add a bit of comment here to explain existence of this plugin snippet?

etanshaul replied:
sure, I can do that

etanshaul replied:
done
